File: python/packages/kagent-adk/src/kagent/adk/_remote_a2a_tool.py
"""Remote A2A agent tool with HITL propagation.

Replaces the upstream ``AgentTool(RemoteA2aAgent(...))`` pairing with a single
``BaseTool`` that directly manages the A2A conversation with a remote agent.

When the remote agent returns ``TaskState.input_required`` (i.e. one of its
tools needs human approval), this tool calls ``request_confirmation()`` to
surface the HITL prompt to the parent agent's flow. On resume the user's
decision is forwarded to the remote agent's pending task.
"""

import logging
import uuid
from typing import Any, Optional
from urllib.parse import urlparse

import httpx
from a2a.client import Client as A2AClient
from a2a.client.card_resolver import A2ACardResolver
from a2a.client.client import ClientConfig as A2AClientConfig
from a2a.client.client_factory import ClientFactory as A2AClientFactory
from a2a.client.errors import A2AClientHTTPError
from a2a.types import (
    AgentCard,
    DataPart,
    Role,
    Task,
    TaskState,
    TextPart,
)
from a2a.types import (
    Message as A2AMessage,
)
from a2a.types import (
    Part as A2APart,
)
from a2a.types import (
    TransportProtocol as A2ATransport,
)
from google.adk.tools.base_tool import BaseTool
from google.adk.tools.tool_context import ToolContext
from google.genai import types as genai_types

from kagent.core.a2a import (
    KAGENT_HITL_DECISION_TYPE_APPROVE,
    KAGENT_HITL_DECISION_TYPE_BATCH,
    KAGENT_HITL_DECISION_TYPE_KEY,
    KAGENT_HITL_DECISION_TYPE_REJECT,
    extract_hitl_info_from_task,
)

logger = logging.getLogger("kagent_adk." + __name__)


def _extract_text_from_task(task: Task) -> str:
    """Extract text content from a completed task's artifacts or status message."""
    # Prefer artifacts (the canonical result)
    if task.artifacts:
        texts: list[str] = []
        for artifact in task.artifacts:
            if artifact.parts:
                for part in artifact.parts:
                    root = part.root if hasattr(part, "root") else part
                    if isinstance(root, TextPart) and root.text:
                        texts.append(root.text)
        if texts:
            return "\n".join(texts)

    # Fall back to status message
    if task.status and task.status.message and task.status.message.parts:
        texts = []
        for part in task.status.message.parts:
            root = part.root if hasattr(part, "root") else part
            if isinstance(root, TextPart) and root.text:
                texts.append(root.text)
        if texts:
            return "\n".join(texts)

    return ""


class KAgentRemoteA2ATool(BaseTool):
    """A tool that calls a remote A2A agent and propagates HITL state.

    Unlike the upstream ``AgentTool(RemoteA2aAgent(...))`` pairing, this tool
    has direct access to ``ToolContext`` and can use ``request_confirmation()``
    to surface subagent HITL prompts to the parent agent's approval flow.
    """

    def __init__(
        self,
        *,
        name: str,
        description: str,
        agent_card_url: str,
        httpx_client: Optional[httpx.AsyncClient] = None,
    ) -> None:
        super().__init__(name=name, description=description)
        self._agent_card_url = agent_card_url
        self._httpx_client = httpx_client
        self._owns_httpx_client = httpx_client is None
        self._a2a_client: Optional[A2AClient] = None
        self._agent_card: Optional[AgentCard] = None
        # Track the context_id from the remote agent for session continuity
        self._last_context_id: Optional[str] = None

    async def close(self) -> None:
        """Close the underlying httpx client if this tool owns it.

        Should be called when the tool is no longer needed to prevent
        resource leaks.  Safe to call multiple times.
        """
        if self._owns_httpx_client and self._httpx_client is not None:
            try:
                await self._httpx_client.aclose()
                logger.debug("Closed httpx client for remote A2A tool %s", self.name)
            except Exception as e:
                logger.warning(
                    "Failed to close httpx client for remote A2A tool %s: %s",
                    self.name,
                    e,
                )
            finally:
                self._httpx_client = None
                self._a2a_client = None

    async def _ensure_client(self) -> A2AClient:
        """Lazily resolve the agent card and initialize the A2A client."""
        if self._a2a_client is not None:
            return self._a2a_client

        # Ensure we have an httpx client
        if self._httpx_client is None:
            self._httpx_client = httpx.AsyncClient(timeout=httpx.Timeout(timeout=600.0))
            self._owns_httpx_client = True

        # Resolve the agent card from URL
        parsed = urlparse(self._agent_card_url)
        base_url = f"{parsed.scheme}://{parsed.netloc}"
        resolver = A2ACardResolver(httpx_client=self._httpx_client, base_url=base_url)
        self._agent_card = await resolver.get_agent_card(relative_card_path=parsed.path)

        if not self._agent_card.url:
            raise ValueError(f"Agent card for {self.name} has no RPC URL")

        # Auto-populate description from agent card if we don't have one
        if not self.description and self._agent_card.description:
            self.description = self._agent_card.description

        # Create the A2A client
        config = A2AClientConfig(
            httpx_client=self._httpx_client,
            streaming=False,
            polling=False,
            supported_transports=[A2ATransport.jsonrpc],
        )
        factory = A2AClientFactory(config=config)
        self._a2a_client = factory.create(self._agent_card)
        return self._a2a_client

    def _get_declaration(self) -> genai_types.FunctionDeclaration:
        """Same schema as AgentTool for agents without an input schema."""
        return genai_types.FunctionDeclaration(
            name=self.name,
            description=self.description,
            parameters=genai_types.Schema(
                type=genai_types.Type.OBJECT,
                properties={
                    "request": genai_types.Schema(type=genai_types.Type.STRING),
                },
                required=["request"],
            ),
        )

    async def run_async(self, *, args: dict[str, Any], tool_context: ToolContext) -> Any:
        """Execute the remote agent tool.

        Phase 1 (first invocation): Send the request to the remote agent.
          - If completed: return the result text.
          - If input_required: call request_confirmation() to pause and
            propagate the HITL prompt to the parent. Store subagent task_id
            and context_id in the confirmation payload.
          - If failed: return the error message.

        Phase 2 (resume after HITL): Forward the user's decision to the
        remote agent's pending task and return the final result.
        """
        if tool_context.tool_confirmation is not None:
            return await self._handle_resume(tool_context)

        return await self._handle_first_call(args, tool_context)

    async def _handle_first_call(self, args: dict[str, Any], tool_context: ToolContext) -> Any:
        """Phase 1: Send the request to the remote agent."""
        client = await self._ensure_client()

        request_text = args.get("request", "")
        message = A2AMessage(
            message_id=str(uuid.uuid4()),
            parts=[A2APart(root=TextPart(text=request_text))],
            role=Role.user,
            # Pass context_id for session continuity with stateful remote agents
            context_id=self._last_context_id,
        )

        task: Optional[Task] = None
        try:
            async for response in client.send_message(request=message):
                if isinstance(response, tuple):
                    # ClientEvent: (Task, UpdateEvent | None)
                    task = response[0]
                elif isinstance(response, A2AMessage):
                    # Direct message response (no task management)
                    if response.context_id:
                        self._last_context_id = response.context_id
                    return self._extract_text_from_message(response)
        except A2AClientHTTPError as e:
            return f"Remote agent '{self.name}' request failed: {e}"
        except Exception as e:
            logger.error("Error calling remote agent %s: %s", self.name, e, exc_info=True)
            return f"Remote agent '{self.name}' request failed: {e}"

        if task is None:
            return f"Remote agent '{self.name}' returned no result."

        # Track context_id for future requests to the same remote agent
        if task.context_id:
            self._last_context_id = task.context_id

        state = task.status.state if task.status else None

        if state == TaskState.input_required:
            return self._handle_input_required(task, tool_context)

        if state == TaskState.failed:
            error_text = _extract_text_from_task(task)
            return error_text or f"Remote agent '{self.name}' failed."

        # completed or any other terminal state
        return _extract_text_from_task(task) or ""

    def _handle_input_required(self, task: Task, tool_context: ToolContext) -> dict[str, Any]:
        """Handle a subagent that returned input_required (HITL).

        Calls request_confirmation() to pause the parent agent and surface
        the HITL prompt to the UI.  The subagent's task_id and context_id
        are stored in the confirmation payload so the resume path can
        forward the user's decision.
        """
        hitl_parts = extract_hitl_info_from_task(task)

        # Build a human-readable hint describing what the subagent is waiting for
        inner_tool_names: list[str] = []
        if hitl_parts:
            for hp in hitl_parts:
                if hp.tool_name:
                    inner_tool_names.append(hp.tool_name)

        if inner_tool_names:
            hint = f"Remote agent '{self.name}' requires approval for tool(s): {', '.join(inner_tool_names)}"
        else:
            hint = f"Remote agent '{self.name}' requires human input before continuing."

        # Serialize HitlPartInfo models to dicts for the payload
        payload = {
            "task_id": task.id,
            "context_id": task.context_id,
            "subagent_name": self.name,
            "hitl_parts": [hp.model_dump(by_alias=True) for hp in hitl_parts] if hitl_parts else None,
        }

        logger.info(
            "Subagent %s returned input_required (task=%s), requesting confirmation from parent",
            self.name,
            task.id,
        )

        tool_context.request_confirmation(hint=hint, payload=payload)
        return {"status": "pending", "waiting_for": "subagent_approval", "subagent": self.name}

    async def _handle_resume(self, tool_context: ToolContext) -> Any:
        """Phase 2: Forward the user's decision to the remote agent's pending task."""
        confirmation = tool_context.tool_confirmation
        payload = confirmation.payload or {}

        logger.info(
            "DEBUG_ASKUSER _handle_resume: confirmed=%s, payload_keys=%s, payload=%s",
            confirmation.confirmed,
            list(payload.keys()) if payload else None,
            payload,
        )

        task_id = payload.get("task_id")
        context_id = payload.get("context_id")
        subagent_name = payload.get("subagent_name", self.name)

        if not task_id:
            logger.error("Resume for %s but no task_id in confirmation payload", self.name)
            return f"Cannot resume remote agent '{subagent_name}': missing task context."

        # Build the decision message.
        # The parent executor merges its own data into the payload alongside
        # the original request_confirmation payload (task_id, context_id, etc).
        # We detect the kind of decision and forward the relevant keys.
        decision_type = None
        batch_decisions = payload.get("batch_decisions")
        ask_user_answers = payload.get("answers")

        if batch_decisions and isinstance(batch_decisions, dict):
            # Per-tool batch decisions (mixed approve/reject for inner tools)
            decision_type = KAGENT_HITL_DECISION_TYPE_BATCH
            decision_data: dict[str, Any] = {
                KAGENT_HITL_DECISION_TYPE_KEY: KAGENT_HITL_DECISION_TYPE_BATCH,
                "decisions": batch_decisions,
            }
            rej_reasons = payload.get("rejection_reasons")
            if rej_reasons and isinstance(rej_reasons, dict):
                decision_data["rejection_reasons"] = rej_reasons
        elif ask_user_answers and isinstance(ask_user_answers, list):
            # ask_user answers — forward as approve + answers so the
            # subagent's _process_hitl_decision takes the ask_user path
            decision_type = KAGENT_HITL_DECISION_TYPE_APPROVE
            decision_data = {
                KAGENT_HITL_DECISION_TYPE_KEY: KAGENT_HITL_DECISION_TYPE_APPROVE,
                "ask_user_answers": ask_user_answers,
            }
            logger.info(
                "DEBUG_ASKUSER _handle_resume: forwarding ask_user_answers to subagent, answers=%s",
                ask_user_answers,
            )
        else:
            if confirmation.confirmed:
                decision_type = KAGENT_HITL_DECISION_TYPE_APPROVE
            else:
                decision_type = KAGENT_HITL_DECISION_TYPE_REJECT
            decision_data = {KAGENT_HITL_DECISION_TYPE_KEY: decision_type}
            # Include rejection reason if available
            if not confirmation.confirmed and payload:
                reason = payload.get("rejection_reason")
                if reason:
                    decision_data["rejection_reason"] = reason

        decision_message = A2AMessage(
            message_id=str(uuid.uuid4()),
            task_id=task_id,
            context_id=context_id,
            role=Role.user,
            parts=[A2APart(root=DataPart(data=decision_data))],
        )

        logger.info(
            "Forwarding %s decision to subagent %s task %s",
            decision_type,
            subagent_name,
            task_id,
        )

        logger.info(
            "DEBUG_ASKUSER _handle_resume: sending decision_message to subagent %s, task_id=%s, decision_data=%s",
            subagent_name,
            task_id,
            decision_data,
        )

        client = await self._ensure_client()
        task: Optional[Task] = None
        try:
            async for response in client.send_message(request=decision_message):
                logger.info(
                    "DEBUG_ASKUSER _handle_resume: got response from subagent, type=%s, response=%s",
                    type(response).__name__,
                    response,
                )
                if isinstance(response, tuple):
                    task = response[0]
                elif isinstance(response, A2AMessage):
                    return self._extract_text_from_message(response)
        except A2AClientHTTPError as e:
            logger.error("DEBUG_ASKUSER _handle_resume: A2AClientHTTPError: %s", e, exc_info=True)
            return f"Remote agent '{subagent_name}' resume failed: {e}"
        except Exception as e:
            logger.error(
                "DEBUG_ASKUSER _handle_resume: Error resuming remote agent %s: %s", subagent_name, e, exc_info=True
            )
            return f"Remote agent '{subagent_name}' resume failed: {e}"

        if task is None:
            logger.warning("DEBUG_ASKUSER _handle_resume: task is None after resume")
            return f"Remote agent '{subagent_name}' returned no result after resume."

        state = task.status.state if task.status else None
        logger.info(
            "DEBUG_ASKUSER _handle_resume: subagent task state after resume: %s, task_id=%s",
            state,
            task.id,
        )

        if state == TaskState.input_required:
            # The subagent has another HITL request (e.g. multiple tools needing
            # approval in sequence). Surface it again.
            return self._handle_input_required(task, tool_context)

        if state == TaskState.failed:
            error_text = _extract_text_from_task(task)
            return error_text or f"Remote agent '{subagent_name}' failed after resume."

        return _extract_text_from_task(task) or ""

    @staticmethod
    def _extract_text_from_message(message: A2AMessage) -> str:
        """Extract text from a direct A2A Message response."""
        if not message.parts:
            return ""
        texts: list[str] = []
        for part in message.parts:
            root = part.root if hasattr(part, "root") else part
            if isinstance(root, TextPart) and root.text:
                texts.append(root.text)
        return "\n".join(texts)

File: go/adk/pkg/tools/remote_a2a_tool.go
package tools

import (
	"context"
	"fmt"
	"log/slog"
	"net/http"
	"strings"

	a2atype "github.com/a2aproject/a2a-go/a2a"
	"github.com/a2aproject/a2a-go/a2aclient"
	"github.com/a2aproject/a2a-go/a2aclient/agentcard"
	"github.com/kagent-dev/kagent/go/adk/pkg/a2a"
	"google.golang.org/adk/agent"
	"google.golang.org/adk/model"
	"google.golang.org/adk/tool"
	"google.golang.org/genai"
)

// SubagentSessionProvider is a tool that delegates to a remote subagent and
// can expose the subagent's session ID for live activity polling.
type SubagentSessionProvider interface {
	tool.Tool
	SubagentSessionID() string
}

// userIDContextKey is the context key for passing the session user_id to the subagent.
type userIDContextKey struct{}

// User ID forwarding interceptor
type userIDForwardingInterceptor struct {
	a2aclient.PassthroughInterceptor
}

func (u *userIDForwardingInterceptor) Before(ctx context.Context, req *a2aclient.Request) (context.Context, error) {
	if uid, ok := ctx.Value(userIDContextKey{}).(string); ok && uid != "" {
		req.Meta.Append("x-user-id", uid)
	}
	return ctx, nil
}

// KAgentRemoteA2ATool calls a remote A2A agent and propagates HITL state.
type KAgentRemoteA2ATool struct {
	name         string
	description  string
	baseURL      string
	httpClient   *http.Client
	extraHeaders map[string]string

	a2aClient *a2aclient.Client
	agentCard *a2atype.AgentCard

	lastContextID string
}

// NewKAgentRemoteA2ATool creates a KAgentRemoteA2ATool.
// baseURL is the base URL of the remote agent (e.g. http://host:port).
// The agent card is fetched lazily from baseURL/.well-known/agent.json.
// If httpClient is nil, http.DefaultClient is used.
func NewKAgentRemoteA2ATool(name, description, baseURL string, httpClient *http.Client, extraHeaders map[string]string) *KAgentRemoteA2ATool {
	if httpClient == nil {
		httpClient = http.DefaultClient
	}
	return &KAgentRemoteA2ATool{
		name:          name,
		description:   description,
		baseURL:       baseURL,
		httpClient:    httpClient,
		extraHeaders:  extraHeaders,
		lastContextID: a2atype.NewContextID(),
	}
}

// Name implements tool.Tool.
func (t *KAgentRemoteA2ATool) Name() string { return t.name }

// Description implements tool.Tool.
func (t *KAgentRemoteA2ATool) Description() string { return t.description }

// IsLongRunning implements tool.Tool.
func (t *KAgentRemoteA2ATool) IsLongRunning() bool { return false }

// SubagentSessionID implements SubagentSessionProvider.
func (t *KAgentRemoteA2ATool) SubagentSessionID() string { return t.lastContextID }

// Declaration returns the GenAI FunctionDeclaration. Mirrors AgentTool schema.
func (t *KAgentRemoteA2ATool) Declaration() *genai.FunctionDeclaration {
	return &genai.FunctionDeclaration{
		Name:        t.name,
		Description: t.description,
		Parameters: &genai.Schema{
			Type: genai.TypeObject,
			Properties: map[string]*genai.Schema{
				"request": {
					Type:        genai.TypeString,
					Description: "The question or task to send to this subagent.",
				},
			},
			Required: []string{"request"},
		},
	}
}

// Run executes the remote agent tool.
func (t *KAgentRemoteA2ATool) Run(ctx tool.Context, args any) (map[string]any, error) {
	if ctx.ToolConfirmation() != nil {
		return t.handleResume(ctx)
	}
	return t.handleFirstCall(ctx, args)
}

// ProcessRequest packs the tool's FunctionDeclaration into the LLM request
// so the model knows it can call this remote A2A agent tool.
func (t *KAgentRemoteA2ATool) ProcessRequest(_ tool.Context, req *model.LLMRequest) error {
	return packTool(req, t.Name(), t.Declaration(), t)
}

// ensureClient lazily resolves the agent card and initialises the A2A client.
func (t *KAgentRemoteA2ATool) ensureClient(ctx context.Context) (*a2aclient.Client, error) {
	if t.a2aClient != nil {
		return t.a2aClient, nil
	}

	resolver := agentcard.NewResolver(t.httpClient)

	// Build ResolveOptions for extra headers.
	var resolveOpts []agentcard.ResolveOption
	for k, v := range t.extraHeaders {
		resolveOpts = append(resolveOpts, agentcard.WithRequestHeader(k, v))
	}

	card, err := resolver.Resolve(ctx, t.baseURL, resolveOpts...)
	if err != nil {
		return nil, fmt.Errorf("failed to resolve agent card for %s: %w", t.name, err)
	}
	t.agentCard = card

	// Auto-populate description from agent card when not explicitly set.
	if t.description == "" && card.Description != "" {
		t.description = card.Description
	}

	opts := []a2aclient.FactoryOption{
		a2aclient.WithJSONRPCTransport(t.httpClient),
	}
	// Always inject x-kagent-source: subagent to mark this as a subagent call.
	meta := a2aclient.CallMeta{}
	meta.Append("x-kagent-source", "agent")
	for k, v := range t.extraHeaders {
		meta.Append(k, v)
	}
	opts = append(opts, a2aclient.WithInterceptors(
		a2aclient.NewStaticCallMetaInjector(meta),
		&userIDForwardingInterceptor{},
	))

	client, err := a2aclient.NewFromCard(ctx, card, opts...)
	if err != nil {
		return nil, fmt.Errorf("failed to create A2A client for %s: %w", t.name, err)
	}
	t.a2aClient = client
	return client, nil
}

// handleFirstCall is Phase 1: send the request to the remote agent.
func (t *KAgentRemoteA2ATool) handleFirstCall(ctx tool.Context, args any) (map[string]any, error) {
	argsMap, _ := args.(map[string]any)
	requestText, _ := argsMap["request"].(string)

	client, err := t.ensureClient(ctx)
	if err != nil {
		return map[string]any{"error": err.Error()}, nil
	}

	message := a2atype.NewMessage(
		a2atype.MessageRoleUser,
		a2atype.TextPart{Text: requestText},
	)
	message.ContextID = t.lastContextID

	// Propagate the session user_id to the subagent via x-user-id header.
	sendCtx := context.WithValue(ctx, userIDContextKey{}, ctx.UserID())
	result, err := client.SendMessage(sendCtx, &a2atype.MessageSendParams{Message: message})
	if err != nil {
		slog.Error("Remote agent request failed", "tool", t.name, "error", err)
		return map[string]any{"error": fmt.Sprintf("Remote agent '%s' request failed: %v", t.name, err)}, nil
	}

	return t.processResult(ctx, result)
}

// handleResume is Phase 2: forward the user's decision to the remote agent's pending task.
func (t *KAgentRemoteA2ATool) handleResume(ctx tool.Context) (map[string]any, error) {
	confirmation := ctx.ToolConfirmation()
	payload, _ := confirmation.Payload.(map[string]any)
	hitlPayload := a2a.ParseHitlConfirmationPayload(payload)

	taskID := hitlPayload.TaskID
	contextID := hitlPayload.ContextID
	subagentName := hitlPayload.SubagentName
	if subagentName == "" {
		subagentName = t.name
	}

	if taskID == "" {
		slog.Error("Resume for remote agent but no task_id in confirmation payload", "tool", t.name)
		return map[string]any{"error": fmt.Sprintf("Cannot resume remote agent '%s': missing task context.", subagentName)}, nil
	}

	// Build the decision DataPart to forward to the subagent.
	decisionData := buildDecisionData(confirmation.Confirmed, hitlPayload)

	message := &a2atype.Message{
		ID:        a2atype.NewMessageID(),
		TaskID:    a2atype.TaskID(taskID),
		ContextID: contextID,
		Role:      a2atype.MessageRoleUser,
		Parts:     a2atype.ContentParts{a2atype.DataPart{Data: decisionData}},
	}

	decisionType, _ := decisionData[a2a.KAgentHitlDecisionTypeKey].(string)
	slog.Info("Forwarding decision to subagent",
		"decisionType", decisionType,
		"subagent", subagentName,
		"taskID", taskID,
	)

	client, err := t.ensureClient(ctx)
	if err != nil {
		return map[string]any{"error": err.Error()}, nil
	}

	// Propagate the session user_id to the subagent via x-user-id header.
	sendCtx := context.WithValue(ctx, userIDContextKey{}, ctx.UserID())
	result, err := client.SendMessage(sendCtx, &a2atype.MessageSendParams{Message: message})
	if err != nil {
		slog.Error("Remote agent resume failed", "tool", subagentName, "error", err)
		return map[string]any{"error": fmt.Sprintf("Remote agent '%s' resume failed: %v", subagentName, err)}, nil
	}

	ret, retErr := t.processResult(ctx, result)
	// Prefer the context_id from the confirmation payload (the original subagent
	// session) over the pre-generated one. Mirrors Python's:
	//   "subagent_session_id": context_id or self._last_context_id
	if retErr == nil && ret != nil {
		sessionID := contextID
		if sessionID == "" {
			sessionID = t.lastContextID
		}
		ret["subagent_session_id"] = sessionID
	}
	return ret, retErr
}

// processResult converts a SendMessageResult into a tool return value,
// handling completed, input_required, and failed states.
func (t *KAgentRemoteA2ATool) processResult(ctx tool.Context, result a2atype.SendMessageResult) (map[string]any, error) {
	switch r := result.(type) {
	case *a2atype.Message:
		return map[string]any{"result": extractTextFromMessage(r)}, nil
	case *a2atype.Task:
		state := r.Status.State
		switch state {
		case a2atype.TaskStateInputRequired:
			return t.handleInputRequired(ctx, r), nil
		case a2atype.TaskStateFailed:
			text := extractTextFromTask(r)
			if text == "" {
				text = fmt.Sprintf("Remote agent '%s' failed.", t.name)
			}
			return map[string]any{"error": text}, nil
		default:
			// completed — include sub-agent's final LLM usage from task.metadata
			// so the parent can display it on the AgentCall card in the UI.
			// Mirrors Python's _extract_usage_from_task(task).
			text := extractTextFromTask(r)
			result := map[string]any{
				"result":              text,
				"subagent_session_id": t.lastContextID,
			}
			if usage := extractUsageFromTask(r); usage != nil {
				result["kagent_usage_metadata"] = usage
			}
			return result, nil
		}
	default:
		return map[string]any{"error": fmt.Sprintf("Remote agent '%s' returned no result.", t.name)}, nil
	}
}

// handleInputRequired pauses parent agent execution via RequestConfirmation,
// storing task_id and context_id so the resume path can forward the decision.
func (t *KAgentRemoteA2ATool) handleInputRequired(ctx tool.Context, task *a2atype.Task) map[string]any {
	if task == nil {
		slog.Error("Subagent returned input_required without task", "tool", t.name)
		return map[string]any{
			"error": fmt.Sprintf("Remote agent '%s' returned input_required without task context.", t.name),
		}
	}

	var hitlParts []a2a.HitlPartInfo
	if task.Status.Message != nil {
		hitlParts = a2a.ExtractHitlInfoFromParts(task.Status.Message.Parts)
	}

	var innerToolNames []string
	for _, hp := range hitlParts {
		if hp.OriginalFunctionCall.Name != "" {
			innerToolNames = append(innerToolNames, hp.OriginalFunctionCall.Name)
		}
	}

	var hint string
	if len(innerToolNames) > 0 {
		hint = fmt.Sprintf("Remote agent '%s' requires approval for tool(s): %s",
			t.name, strings.Join(innerToolNames, ", "))
	} else {
		hint = fmt.Sprintf("Remote agent '%s' requires human input before continuing.", t.name)
	}

	confirmPayload := a2a.HitlConfirmationPayload{
		TaskID:       string(task.ID),
		ContextID:    task.ContextID,
		SubagentName: t.name,
		HitlParts:    hitlParts,
	}

	slog.Info("Subagent returned input_required, requesting confirmation from parent",
		"tool", t.name, "taskID", task.ID)

	if err := ctx.RequestConfirmation(hint, confirmPayload.ToMap()); err != nil {
		slog.Error("Failed to request confirmation", "tool", t.name, "error", err)
	}
	return map[string]any{
		"status":      "pending",
		"waiting_for": "subagent_approval",
		"subagent":    t.name,
	}
}

// buildDecisionData constructs the decision DataPart.Data map to forward to the subagent.
func buildDecisionData(confirmed bool, payload a2a.HitlConfirmationPayload) map[string]any {
	switch {
	case len(payload.BatchDecisions) > 0:
		batchDecisions := make(map[string]any, len(payload.BatchDecisions))
		for id, decision := range payload.BatchDecisions {
			batchDecisions[id] = string(decision)
		}
		data := map[string]any{
			a2a.KAgentHitlDecisionTypeKey: a2a.KAgentHitlDecisionTypeBatch,
			a2a.KAgentHitlDecisionsKey:    batchDecisions,
		}
		if len(payload.RejectionReasons) > 0 {
			rejReasons := make(map[string]any, len(payload.RejectionReasons))
			for id, reason := range payload.RejectionReasons {
				rejReasons[id] = reason
			}
			data[a2a.KAgentHitlRejectionReasonsKey] = rejReasons
		}
		return data

	case len(payload.Answers) > 0:
		askUserAnswers := make([]map[string]any, 0, len(payload.Answers))
		for _, answer := range payload.Answers {
			askUserAnswers = append(askUserAnswers, map[string]any{"answer": answer.Answer})
		}
		// Forward as approve + answers so the subagent's HITL processor takes the ask_user path.
		return map[string]any{
			a2a.KAgentHitlDecisionTypeKey: a2a.KAgentHitlDecisionTypeApprove,
			a2a.KAgentAskUserAnswersKey:   askUserAnswers,
		}

	default:
		decisionType := a2a.KAgentHitlDecisionTypeApprove
		if !confirmed {
			decisionType = a2a.KAgentHitlDecisionTypeReject
		}
		data := map[string]any{a2a.KAgentHitlDecisionTypeKey: decisionType}
		if !confirmed {
			if payload.RejectionReason != "" {
				data["rejection_reason"] = payload.RejectionReason
			}
		}
		return data
	}
}

// extractUsageFromTask extracts kagent_usage_metadata from a completed task.
// The A2A task manager merges the final TaskStatusUpdateEvent.metadata into
// task.metadata, so LLM usage is available here for non-streaming callers.
// Port of _remote_a2a_tool.py:_extract_usage_from_task().
func extractUsageFromTask(task *a2atype.Task) map[string]any {
	if task == nil || task.Metadata == nil {
		return nil
	}
	usage, ok := task.Metadata["kagent_usage_metadata"].(map[string]any)
	if ok && len(usage) > 0 {
		return usage
	}
	return nil
}

// extractTextFromTask extracts the text result from a completed Task.
func extractTextFromTask(task *a2atype.Task) string {
	if task == nil {
		return ""
	}
	// Prefer artifacts (canonical result).
	if len(task.Artifacts) > 0 {
		var texts []string
		for _, artifact := range task.Artifacts {
			for _, part := range artifact.Parts {
				if tp, ok := part.(a2atype.TextPart); ok && tp.Text != "" {
					texts = append(texts, tp.Text)
				}
			}
		}
		if len(texts) > 0 {
			return strings.Join(texts, "\n")
		}
	}
	// Fall back to status message.
	if task.Status.Message != nil {
		return extractTextFromMessage(task.Status.Message)
	}
	return ""
}

// extractTextFromMessage extracts text from a direct A2A Message response.
func extractTextFromMessage(message *a2atype.Message) string {
	if message == nil {
		return ""
	}
	var texts []string
	for _, part := range message.Parts {
		if tp, ok := part.(a2atype.TextPart); ok && tp.Text != "" {
			texts = append(texts, tp.Text)
		}
	}
	return strings.Join(texts, "\n")
}

// KAgentRemoteA2AToolset wraps KAgentRemoteA2ATool as an ADK Toolset so the
// runner's close path can destroy the underlying A2A client.
type KAgentRemoteA2AToolset struct {
	tool *KAgentRemoteA2ATool
}

// NewKAgentRemoteA2AToolset creates a KAgentRemoteA2AToolset.
func NewKAgentRemoteA2AToolset(name, description, baseURL string, httpClient *http.Client, extraHeaders map[string]string) *KAgentRemoteA2AToolset {
	return &KAgentRemoteA2AToolset{
		tool: NewKAgentRemoteA2ATool(name, description, baseURL, httpClient, extraHeaders),
	}
}

// Name returns the toolset name (== inner tool name).
func (s *KAgentRemoteA2AToolset) Name() string { return s.tool.name }

// SubagentSessionID implements SubagentSessionProvider at the toolset level.
func (s *KAgentRemoteA2AToolset) SubagentSessionID() string { return s.tool.SubagentSessionID() }

// Tools implements tool.Toolset.
func (s *KAgentRemoteA2AToolset) Tools(_ agent.ReadonlyContext) ([]tool.Tool, error) {
	return []tool.Tool{s.tool}, nil
}

// Close destroys the underlying A2A client, releasing any connections.
func (s *KAgentRemoteA2AToolset) Close() error {
	if s.tool.a2aClient != nil {
		if err := s.tool.a2aClient.Destroy(); err != nil {
			slog.Warn("Failed to destroy A2A client", "tool", s.tool.name, "error", err)
		}
		s.tool.a2aClient = nil
	}
	return nil
}
